src/app/authoring-tool/edit-dynamic-prompt/edit-dynamic-prompt.component.ts

```diff
@@ -0,0 +1,46 @@
+import { Component, EventEmitter, Input, OnInit, Output } from '@angular/core';
+import { DynamicPrompt } from '../../../assets/wise5/directives/dynamic-prompt/DynamicPrompt';
+import { ProjectService } from '../../../assets/wise5/services/projectService';
+
+@Component({
+  selector: 'edit-dynamic-prompt',
+  templateUrl: './edit-dynamic-prompt.component.html',
+  styleUrls: ['./edit-dynamic-prompt.component.scss']
+})
+export class EditDynamicPromptComponent implements OnInit {
+  allowedConnectedComponentTypes: string[] = ['OpenResponse'];
+  @Input() authoringComponentContent: any;
+  @Output() dynamicPromptChangedEvent = new EventEmitter<string>();
+
+  constructor(private projectService: ProjectService) {}
+
+  ngOnInit(): void {}
+
+  toggleDynamicPrompt(event: any): void {
+    if (this.authoringComponentContent.dynamicPrompt == null) {
+      this.authoringComponentContent.dynamicPrompt = new DynamicPrompt({
+        referenceComponent: {},
+        rules: []
+      });
+    }
+    this.authoringComponentContent.dynamicPrompt.enabled = event.checked;
+    this.dynamicPromptChangedEvent.next(event);
+  }
+
+  referenceComponentNodeIdChanged(event: any): void {
+    const dynamicPrompt = this.authoringComponentContent.dynamicPrompt;
+    dynamicPrompt.referenceComponent.componentId = null;
+    let numberOfAllowedComponents = 0;
+    let allowedComponent = null;
+    for (const component of this.projectService.getComponentsByNodeId(event.nodeId)) {
+      if (this.allowedConnectedComponentTypes.includes(component.type)) {
+        numberOfAllowedComponents += 1;
+        allowedComponent = component;
+      }
+    }
+    if (numberOfAllowedComponents === 1) {
+      dynamicPrompt.referenceComponent.componentId = allowedComponent.id;
+    }
+    this.dynamicPromptChangedEvent.next(event);
```

Comment:
Suggestions: 
- Move variable assignment together
- Only call ```dynamicPrompt.referenceComponent.componentId = null;``` when necessary instead of every time this function is called
- Use shorter variable name: numberOfAllowedComponents -> numAllowedComponents
- Rename variable: allowedConnectedComponentTypes -> allowedReferenceComponentTypes
```suggestion
    let numAllowedComponents = 0;
    let allowedComponent = null;
    for (const component of this.projectService.getComponentsByNodeId(event.nodeId)) {
      if (this.allowedReferenceComponentTypes.includes(component.type)) {
        numAllowedComponents += 1;
        allowedComponent = component;
      }
    }
    const dynamicPrompt = this.authoringComponentContent.dynamicPrompt;
    if (numAllowedComponents === 1) {
      dynamicPrompt.referenceComponent.componentId = allowedComponent.id;
    } else {
      dynamicPrompt.referenceComponent.componentId = null;
    }
    this.dynamicPromptChangedEvent.next(event);
```